### PR TITLE
2nd attempt - Windows support, Alpha release

### PR DESCRIPTION
**Pull request by sergiimatusEPAM**
SM : fixed all showstoppers at https://github.com/dcos/dcos-ansible/pull/37

### CODE REVIEW

File: dcos.yml
Comment (fatz) [inline dcos.yml]: i think we should also wait for windows agents

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (fatz) [inline roles/DCOS.agent_windows/tasks/main.yml]: should be `dcos['version']` which should be always specified

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (fatz) [inline roles/DCOS.agent_windows/tasks/main.yml]: oh i've overseen the `delegate_to` can't we just do this on the windows node?

Comment (fatz) [reply roles/DCOS.agent_windows/tasks/main.yml]: https://docs.ansible.com/ansible/latest/modules/win_get_url_module.html would be a sufficient alternative?

Comment (fatz) [reply roles/DCOS.agent_windows/tasks/main.yml]: otherwise the /dcos might be problematic for unprivileged users. I thought this one is on the windows node

File: roles/DCOS.agent_windows/vars/main.yml
Comment (fatz) [inline roles/DCOS.agent_windows/vars/main.yml]: as mentioned in the set_facts part. The var is `dcos['version']`

**Review by fatz — CHANGES_REQUESTED** [review]
looks good beside some minor things I found.

Please use the `dcos['version']` which is holding the version to install. Hint if you want to only support from a given dcos version on you could use a statement like `when: "dcos['version'] is version('1.9', '>=')"` 

Beside this I just realized the download part was using `delegate_to: localhost` this should be avoided as the path will probably add issues to unprivileged users.

Better download directly on the agent.

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: ansible_password needs to be passed to WinRM as it fails to auth. I've added "Wait for WIndows Instances .." right after Linux routines

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml]: I see , agreed. But as of now our changes are not in your docker image. For Alpha, it was agreed to use git clone. Regardless, similar mesosphere_universal_installer_dcos.yml should be used by windows part now.

for now, in local vars I've added stub with `dcos: version: 1.13` 
Will fix once merged.

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml]: yes, I was thinking of it before, but now added as well

File: roles/DCOS.agent_windows/vars/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/vars/main.yml]: yup

**Review by sergiimatusEPAM — COMMENTED** [review]
@fatz , Please have a look this one. I've tested on AWS DEV1 env, so should be fine..
Once a new DOcker image built it will be easy to manage version/commit , and can make upgrade tasks

File: dcos.yml
Comment (fatz) [reply dcos.yml]: but sholdn't this be given by the inventory file?

Comment (fatz) [reply dcos.yml]: just checked this is simply done by the correct var in inventory

File: dcos.yml
Comment (fatz) [inline dcos.yml]: please remove this line. We should still gather facts here

File: dcos.yml
Comment (fatz) [inline dcos.yml]: no need for a nother newline

File: dcos.yml
Comment (fatz) [inline dcos.yml]: no need for a nother newline

File: dcos.yml
Comment (fatz) [inline dcos.yml]: no need for a nother newline

File: dcos.yml
Comment (fatz) [inline dcos.yml]: one newline is enough

File: dcos.yml
Comment (fatz) [inline dcos.yml]: this part is not needed we already gather facts in the beginning

File: inventory.example
Comment (fatz) [inline inventory.example]: `agent1-windows.example.com ansible_user=Administrator ansible_password=<mysecurepassword>`

File: inventory.example
Comment (fatz) [inline inventory.example:53]: to support proper winrm transport
```
[agents_windows:vars]
ansible_connection=winrm
ansible_winrm_transport=basic
ansible_winrm_server_cert_validation=ignore
```

File: inventory.example
Comment (fatz) [inline inventory.example]: rename to `agents_windows` to match the same naming as other agents

File: README.md
Comment (fatz) [inline README.md]: `agents_windows` is a better name

File: README.md
Comment (fatz) [inline README.md:35]: see other comment 
```
[agents_windows:vars]
ansible_connection=winrm
ansible_winrm_transport=basic
ansible_winrm_server_cert_validation=ignore
```

File: README.md
Comment (fatz) [inline README.md]: `agent1-windows.example.com ansible_user=Administrator ansible_password=<mysecurepassword>`

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (fatz) [inline roles/DCOS.agent_windows/tasks/main.yml]: its called `ansible_password` not `pass`

File: roles/DCOS.agent_windows/vars/main.yml
Comment (fatz) [inline roles/DCOS.agent_windows/vars/main.yml]: this is not a valid DC/OS version. either `1.13.0` or `1.13.1`

File: README.md
Comment (fatz) [inline README.md]: why having a blank line here

File: dcos.yml
Comment (fatz) [inline dcos.yml]: why not gathering fact here?

File: dcos.yml
Comment (fatz) [inline dcos.yml]: we should run the setup step for `agents_windows` so please leave it to all

**Review by fatz — CHANGES_REQUESTED** [review]
see my comments.

Also i've tested the module and it restarts the node on every ansible run. Looks like `TASK [DCOS.agent_windows : Windows | Download mesos-binaries]` running every time

**Comment by DominikDary** [comment]
> see my comments.
> 
> Also i've tested the module and it restarts the node on every ansible run. Looks like `TASK [DCOS.agent_windows : Windows | Download mesos-binaries]` running every time

So your suggestion is @fatz to avoid the re-download tasks in every run of change?
From our offline conversation I understood that This leads to a handler notify which leads to unnecessary restart. 

Just rephrasing here to make sure we have alignment and shared understanding.

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (fatz) [inline roles/DCOS.agent_windows/tasks/main.yml:58]: why reboot here? Is this really needed? Other options?

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (fatz) [inline roles/DCOS.agent_windows/tasks/main.yml:4]: isn't this the part where we notify for reboot?

**Review by fatz — CHANGES_REQUESTED** [review]
Just added a some more findings. Please check the handler and also check re running the ansible playbook. A second run without any change should always leed to no change all ok

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml:4]: it is, but we also need to notify reboot once a new win_sheduled_task added. So for now adding there

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml:58]: really, because wants the scheduled task to be run

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: done

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: added gath facts back

File: README.md
Comment (sergiimatusEPAM) [reply README.md]: removed

File: roles/DCOS.agent_windows/vars/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/vars/main.yml]: corrected execution to use -e @/dcos.yml

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml]: agreed, getting this from inventory

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: agreed, I used to do this from Python - 2 new lines

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: agreed

File: inventory.example
Comment (sergiimatusEPAM) [reply inventory.example]: agreed

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: agreed

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: agreed

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: agreed, removed

File: inventory.example
Comment (sergiimatusEPAM) [reply inventory.example]: corrected

File: inventory.example
Comment (sergiimatusEPAM) [reply inventory.example:53]: added

File: README.md
Comment (sergiimatusEPAM) [reply README.md]: corrected to `agent_windows`

File: README.md
Comment (sergiimatusEPAM) [reply README.md:35]: added

File: README.md
Comment (sergiimatusEPAM) [reply README.md]: added to inventory example

**Comment by sergiimatusEPAM** [comment]
Will 

> > see my comments.
> > Also i've tested the module and it restarts the node on every ansible run. Looks like `TASK [DCOS.agent_windows : Windows | Download mesos-binaries]` running every time
> 
> So your suggestion is @fatz to avoid the re-download tasks in every run of change?
> From our offline conversation I understood that This leads to a handler notify which leads to unnecessary restart.
> 
> Just rephrasing here to make sure we have alignment and shared understanding.

Fixed this part as well.

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml:58]: Yes, reboot requires. But be aware that the reboot will happen only if the new binaries unpacked and subsequently if new scheduled job/service created. All needed changes were performed

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml:4]: in fact this one should be , but if we create a new scheduled job, it also needs to happen. INstead of 2 reboots, we ended up with 1 after Scheduled Job creation.

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: changed to agents_windows

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: modified to gathering

File: dcos.yml
Comment (sergiimatusEPAM) [reply dcos.yml]: ansible_password is now passed by inventory. the issue with gath facts gone

**Comment by sergiimatusEPAM** [comment]
@sebbrandt87 , please review and merge in @fatz absense

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sebbrandt87) [inline roles/DCOS.agent_windows/tasks/main.yml:16]: this can be even more easy without an extra check for the present zip: 
see https://docs.ansible.com/ansible/latest/modules/win_get_url_module.html, option `force` ;)

File: Dockerfile
Comment (sebbrandt87) [inline Dockerfile]: extra space?

File: README.md
Comment (sebbrandt87) [inline README.md]: but still one extra empty line
please remove

File: inventory.example
Comment (sebbrandt87) [inline inventory.example:46]: I think we should have this part also in the README.md

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sebbrandt87) [inline roles/DCOS.agent_windows/tasks/main.yml]: extra space after `arguments`

File: README.md
Comment (sebbrandt87) [inline README.md]: please name `agents_windows` as it is the used identifier later

**Review by sebbrandt87 — CHANGES_REQUESTED** [review]

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml:16]: yeah, I see it will work . But it's a bit incorrect to set force =Yes by default , IMHO

Comment (sebbrandt87) [reply roles/DCOS.agent_windows/tasks/main.yml:16]: you did not read it correct ;)
the default is `force = Yes | True | true` already. You need to set it to `No` and then we can get rid of the extra action.

File: Dockerfile
Comment (sergiimatusEPAM) [reply Dockerfile]: yes, removed

File: README.md
Comment (sergiimatusEPAM) [reply README.md]: edited

File: README.md
Comment (sergiimatusEPAM) [reply README.md]: removed

File: inventory.example
Comment (sergiimatusEPAM) [reply inventory.example:46]: added

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml]: removed

**Review by sergiimatusEPAM — COMMENTED** [review]
@sebbrandt87 , performed the changes you've asked. Please have a look if you OK now

File: roles/DCOS.agent_windows/tasks/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml:16]: I read it correctly . What I was trying to say is - `win_get_url` is designed incorrectly, because `force` flag set to yes by default. It should be set to no by default instead

Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/tasks/main.yml:16]: however, tested and fixed as per your recommendation

Comment (sebbrandt87) [reply roles/DCOS.agent_windows/tasks/main.yml:16]: ah okay :) so I read your comment wrong ^^ sry

File: README.md
Comment (sebbrandt87) [inline README.md:32]: duplicate

File: roles/DCOS.agent_windows/handlers/main.yml
Comment (sebbrandt87) [inline roles/DCOS.agent_windows/handlers/main.yml]: Why quoting here and within all other steps not?

File: roles/DCOS.agent_windows/vars/main.yml
Comment (sebbrandt87) [inline roles/DCOS.agent_windows/vars/main.yml]: not used anymore if the `url_to_mesosbin` is changed

File: roles/DCOS.agent_windows/vars/main.yml
Comment (sebbrandt87) [inline roles/DCOS.agent_windows/vars/main.yml]: replace `dcos_version_specifier` with `dcos['version']`

**Review by sebbrandt87 — CHANGES_REQUESTED** [review]

File: README.md
Comment (sergiimatusEPAM) [reply README.md]: Sorry, I didn't get you. There is inventory example and single use of `[agents_windows]` group.

File: roles/DCOS.agent_windows/handlers/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/handlers/main.yml]: Removed the handler

File: roles/DCOS.agent_windows/vars/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/vars/main.yml]: removed the variable `dcos_version_specifier` and using `{{ dcos['version'] }}` going forward

File: roles/DCOS.agent_windows/vars/main.yml
Comment (sergiimatusEPAM) [reply roles/DCOS.agent_windows/vars/main.yml]: removed the variable `dcos_version_specifier` and using `{{ dcos['version'] }}` going forward

File: README.md
Comment (sergiimatusEPAM) [reply README.md:32]: sorry, found this duplication now. Removed

**Review by sebbrandt87 — APPROVED** [review]

**Pull request merged**
